hexrdgui/mask_manager.py

```diff
@@ -0,0 +1,281 @@
+
+    def update_mask_array(self):
+        # TODO: rename apply_threshold_mask since its purpose has changed now?
+        # FIXME: Function parameters changed
+        self._masked_arrays = apply_threshold_mask(self.values)
+
+    def serialize(self):
+        return {
+            'min_val': self.min_val,
+            'max_val': self.max_val,
+            'name': self.name,
+            'mtype': self.mask_type,
+            'visible': self.mask_image,
+            'border': self.show_border,
+        }
+
+    def deserialize(self, data):
+        self.name = data['name']
+        self.mask_type = data['mtype']
+        self.mask_image = data['visible']
+        self.show_border = data['border']
+        self.set_data([data['min_val'], data['max_val']])
+
+
+class MaskManager(metaclass=Singleton):
+    def __init__(self, view_mode):
+        self.masks = {}
+        self.view_mode = view_mode
+
+    @property
+    def visible_masks(self):
+        return [k for k, v in self.masks if v.mask_image]
+
+    @property
+    def visible_boundaries(self):
+        return [k for k, v in self.masks if v.show_border]
+
+    @property
+    def threshold_mask(self):
+        for mask in self.masks.values():
+            if mask.mask_type == 'threshold':
+                return mask
+        return None
+
+    def add_mask(self, name, data, mtype, mask_image=True, show_border=False):
+        # Enforce name uniqueness
+        name = unique_name(self.masks.keys(), name)
+        if mtype == 'threshold':
+            new_mask = ThresholdMask(name, mtype, mask_image)
+        else:
+            new_mask = RegionMask(name, mtype, mask_image, show_border)
+        new_mask.set_data(self.view_mode, data)
+        self.masks[name] = new_mask
+
+    def remove_mask(self, name):
+        self.masks.pop(name)
+
+    def write_all_masks(self, h5py_group=None):
+        d = {}
+        for name, mask_info in self.masks:
+            d[name] = mask_info.serialize()
+        if h5py_group:
+            self.write_masks_to_group(d, h5py_group)
+        else:
+            self.export_masks_to_file(d)
+
+    def save_state(self, h5py_group):
+        if 'masks' not in h5py_group:
+            h5py_group.create_group('masks')
+
+        self.write_all_masks(h5py_group['masks'])
+
+    def load_masks(self, h5py_group):
+        # TODO: Handle case of detector name mismatch (loading wrong mask file)
+        items = h5py_group.items()
+        if '_visible' in h5py_group.values():
+            # This is a file using the old format
+            items = load_old_mask_file(h5py_group)
+        actual_view_mode = self.view_mode
+        self.view_mode = ViewType.raw
+        for key, data in items:
+            if data['mtype'] == 'threshold':
+                new_mask = ThresholdMask(None, None)
+                new_mask.deserialize(data)
+            else:
+                new_mask = RegionMask(None, None)
+                new_mask.deserialize(data)
```

Comment:
It might be worth considering making `deserialize` a `classmethod` instead, so that you can just do this:

```python
new_mask = RegionMask.deserialize(data)
```